### DISCUSSION TITLE
CodeProject on AI TOOL?

### DISCUSSION THREAD

**Discussion by Yonny24** [VorlonCD/bi-aidetection#328] [General]
A lot of people appear to be moving away from Deepstack to CP.
Is there going to be an option to add the codeproject server on the AITool in the near future? Or maybe it exists now and I haven't figured it out.
From the drop down I only see deepstack, AWS and some other.

Many thanks

**Comment by Yonny24** [comment]
Appear to have gotten it working fine 32168 being the port of the CP

![image](https://user-images.githubusercontent.com/64428686/216682328-f8516ed8-fd40-40bd-a1ad-89b7e1c995ae.png)

**Comment by mattbruman** [comment]
CP works you just enter the server info, I am running both and CP is used to verify

**Reply by Yonny24** [reply]
> CP works you just enter the server info, I am running both and CP is used to verify

why both if I may ask? Are you not happy with the CP development yet? Hence the verification server option?

**Reply by mattbruman** [reply]
I use the CP for vehicle identification and as the verification for people as I feel I get way less false positives when DS is like 70% sure it goes to CP to be absolutely sure. Also I did as you did had both to test and just never went back.

**Comment by VorlonCD** [comment]
yup, seems to be drop in compatible!   Except of course for the Deepstack tab that controls stopping/starting, etc.

**Comment by Yonny24** [comment]
Nice one - going to install CP on the main server today - was using a test server without deepstack.
Now how do I get it to use GPU like the deepstack? Will it automatically detect the nvidia gpu card?

**Comment by VorlonCD** [comment]
it just automatically uses it I believe.   Just go to its web interface to verify or change that setting:  http://127.0.0.1:32168/

**Comment by Yonny24** [comment]
Sweet!!

![image](https://user-images.githubusercontent.com/64428686/216689540-7243e871-b555-4957-8d7a-986ba2ef976b.png)

**Comment by Yonny24** [comment]
All installed and working.
But I'm finding the CP a little slower aounrd 135ms compare to 90 or so with deepstack (HIGH mode).
It's using YOLOv5 6.2 by default

**Comment by Yonny24** [comment]
got it down to 90ms by adjusting the jpeg quality to 50% under when triggered

**Comment by Yonny24** [comment]
CP installed and Deepstack stopped for now to see how it goes.
Using the AITool.

Now, one section of the AITool I never knew which to configure correctly is the linked server and combine results options.
I use an animal custom server from sunset-sunrise and the normal object one 24hrs a day.

I have tried both options, one as linked server and waits for the results from the other and none of these options ticked.
What is the difference? Just one queues and waits for the other to finish first? If nothing is ticked and both are running what difference will that make?
Thanks!

![image](https://user-images.githubusercontent.com/64428686/216704326-713757a3-0508-4cfa-acd0-d91dde158345.png)

![image](https://user-images.githubusercontent.com/64428686/216704350-2a44ad78-af09-4fe7-b093-c3cec0d04d1a.png)

**Reply by VorlonCD** [reply]
custom would need only "use only as linked server"
deepstack would need "link" enabled with "custom" selected in the dropdown

It does of course add to total processing time since it waits for both to be done in order to combine results.

**Comment by Yonny24** [comment]
having some issues here.
If I tick the custom as linked server and then the main CP standard model drop down to the custom model I am being bombarded with these error messages:

ERROR> Skipping image because queue (101) is greater than '100'. (Adjust 'MaxImageQueueSize' in .JSON file if needed): d:\aiinput\Front.20230206_222451386.jpg

I have disabled the custom CP model for now and just using the default CP model.

What do I need to adjust to avoid these errors? I've tried unticking all queue/queued actions and also enabling them. Same result

Thanks

**Reply by balucanb** [reply]
Same issue here.

**Comment by VorlonCD** [comment]
If you go over to History tab and look at "AnalysisDurationMS" column, are things taking a much longer time after enabling linking?  It would make sense that it might double the time because it has to wait for output from 2 different server results.

One solution is to go into BI and reduce the number of images generated one way or another.    Dont create JPEGS for as many seconds and/or reduce the detection sensitivity.   You are likely overwhelming AITOOL with so many images it cant keep up.

Another solution is to install CodeProject on a second machine (Like I do) and add that as a secondary server.  (Linked the same way).   That way, when the first server is busy it will offload the work to the second server and the queue will stay lower.

**Reply by Yonny24** [reply]
This custom model is taking about 200ms, standard one about 85-90ms.
When I used the same set up with deepstack servers with a custom model linked also, firing off many jpegs from BI I never encountered this issue.

I have a gpu gtx970 and it can fire off as fast as 40ms, super fast. So now the AI tool is being overwhelmed? Nothing has been change on my BI just swapped over to CP- all fine with 2 deepstack running as linked.

I guess I could have the custom model on the BI integration and run the standard on the AITool if I can't find a resolution. Maybe a setting was changing on the AItooll without me knowing about it

Thanks for the updates!

**Comment by VorlonCD** [comment]
I just set it up the same way with CodeProject AI:

1) main /detection is set to "Link" to ipcam-animal  (nothing else checked)
2) custom /ipcam-animal set to "Use ONLY as linked server".

My old average was near 250 and my new is near 600.   Thats with GPU acceleration+GTX 1080TI.   Its a little slow, but BI is also on the same machine using the GPU for video encoding and I record 24 hours so I expect it could be a lot faster if that wasnt running.

No trouble.  Except the AI is hallucinating squirrels everywhere.  

I have 6 cameras.   The stats on the bar at the bottom are 3 history items a minute, 3 images a minute, queue: 2 max.

I have 3 servers, queue is unchecked so it always uses the first CodeProject server.  Second is Deepstack on the same machine and third is on a NAS box but it hardly ever has to use that one unless there is a lot of activity.

In BI I'm creating jpegs for 4 seconds on each movement.  You could try reducing this setting and maybe reduce the sensitivity on the Trigger tab.   If not using a fast NVME drive that would certainly help. 

When it builds up are you getting any errors?  I'm assuming you are just throwing too many images at it and it cant keep up.

**Reply by Yonny24** [reply]
Thanks let me have a play around with it later. I think I have the same 4 sec setting.
I have this on each cam

![image](https://user-images.githubusercontent.com/64428686/217626550-8c87ea4c-dcd5-4395-8d26-c86546c95d0f.png)

Could this max time to wait be an issue - never knew what to put in there.
Also I'm never sure whether to queue items on the main server settings and in each camera for notifications.

![image](https://user-images.githubusercontent.com/64428686/217626940-692d660e-32dc-4bd2-95ce-a4c2d80b9745.png)

Thanks!

**Reply by VorlonCD** [reply]
maybe leave it at 1000 and enable the Error checkbox so you know when it takes longer than that to process.  If it ever errors out, something may be wrong with running two CP instances at once.  Maybe video card memory is too full?   The CP web interface or task manager should tell you.   It is a memory hog and 4 gb is not enough on the card.  I think mine has 12.

**Comment by Yonny24** [comment]
I have set both custom and main wiithout any links, individual, just walked around with my laptop and seems to be working fine.
90-100ms for both servers each image, no errors. 
I did get detected as a dog in a couple of the animal custom detections!! Maybe I need a shave and haircur :)
lol

Is deepstack really disappearing? This is why I got forward with CP just in case

**Comment by Yonny24** [comment]
one doubt why does the coolddown time for telegram always default to 2 seconds? I want it faster but it keeps changing to 2 even though the cameras have 1

**Reply by VorlonCD** [reply]
No work has been done on deepsack for quite some time.

I seem to recall that is a minimum time for the the Telegram server or else it will start failing to upload if you send them too fast.

**Reply by Yonny24** [reply]
Thanks - Yes I saw the download link for some time has not been working. Hence my migration to CP for now.
Never had issues with deepstack though. It worked well with your tool and was fast even on HIGH mode
Can CP be adjusted like deepstack? different modes etc do you know?

**Comment by Yonny24** [comment]
Just an update - all working great for a week now, with both codeproject custom(animal) and default but with no linked server ticked, just separate. seems to work better and faster. Not sure why. 70-100ms each frame.
Thanks!
